### DISCUSSION TITLE
tables support

### DISCUSSION THREAD

**Discussion by alexisargyris** [vkbo/novelWriter#556] [Ideas]
Hi,

Is there any possibility for tables support? Nothing too fancy just simple rows and columns. I need it for a plot summary that takes place in the present and in the past.

**Comment by vkbo** [comment]
Rendering tables is technically possible by the widget used for the document viewer, but unless you use a fixed width font in the editor, it is difficult to generate a table with markdown syntax that is readable in both panels.

I do see the value in having simple tables in the notes, and I do make them myself. But you can achieve simple tables by just using the tab key and hard line breaks. A while back I added a hack that ensures that tabs are properly rendered also in the document viewer (HTML generally doesn't support tabs), so aligning text with tabs is thus readable in both the editor and the viewer.

Hard line breaks in markdown are achieved by leaving two spaces at the end of the line, which novelWriter also supports. This is needed to separate rows in such a table.

Example with editor on the left and viewer on the right:
![image](https://user-images.githubusercontent.com/1619840/104132417-742c2680-537d-11eb-8dfa-c5ac42494a8e.png)

**Comment by alexisargyris** [comment]
Yes, this is quite usable. Unfortunately, I happen to need something like this (i.e. rows of unequal height).

![image](https://user-images.githubusercontent.com/1640855/104153839-0e2ab800-53ec-11eb-9ad9-a2ce94e02980.png)

**Comment by vkbo** [comment]
I see. Well, you can still do this with tabs, with the benefit of having a readable table in both editor and viewer with a variable width font. While writing a parser for markdown tables, especially in a simplified form, the recursive parsing nature of a table doesn't fit with the parsing pipeline used in novelWriter, so it will have to be rewritten. That is a fairly big job.

The parser is written in a linear way because I decided early on not to include complex structures like tables, and stick to minimal formatting features. This makes the parser both faster and easier to subclass to produce multiple output formats. Full markdown parsers are notoriously convoluted due to the nature of the markdown spec. I felt this wasn't needed for writing fiction.

I suppose I could add a simplified table support where the `|` character wrapping text is converted to table cells after the text in the cells have been parsed. That may still fit within a linear parsing algorithm by adding a second step that adds the table boundaries, but I'd have to try it out. I do something similar to apply document layouts and header formatting on exports.

Still, I'm concerned about the readability of this. These kind of data structures are generally better suited for spread sheet type documents. It may actually be a better idea to add a new type of document that can be added to a project. For instance a table document. No formula support, just text in cells. Qt5 already has such a widget, so it's just a matter of defining a format to save the data and adding an editor window for it. Another alternative is to move forward with the earlier idea of allowing attaching external files to notes.

**Reply by alexisargyris** [reply]
Either of these two solutions (i.e. new doc type or external file) works for me.

**Reply by vkbo** [reply]
Great! I'm probably going to support attachments anyway, as it helps connect research to notes, so this is probably the simplest solution.
